src/sync/impl/sync_metadata.cpp

```diff
@@ -358,60 +371,81 @@ util::Optional<std::string> SyncUserMetadata::user_token() const
 
 std::string SyncUserMetadata::auth_server_url() const
 {
+    checkValidOrThrow("auth_server_url");
     REALM_ASSERT(m_realm);
     m_realm->verify_thread();
     return m_row.get_string(m_schema.idx_auth_server_url);
 }
 
 bool SyncUserMetadata::is_admin() const
 {
+    checkValidOrThrow("is_admin");
     REALM_ASSERT(m_realm);
     m_realm->verify_thread();
     return m_row.get_bool(m_schema.idx_user_is_admin);
 }
 
 void SyncUserMetadata::set_user_token(util::Optional<std::string> user_token)
 {
-    if (m_invalid)
+    if (!m_row.is_attached())
         return;
 
     REALM_ASSERT_DEBUG(m_realm);
     m_realm->verify_thread();
     m_realm->begin_transaction();
+    if (!m_row.is_attached()) {
+        m_realm->cancel_transaction();
+        return;
+    }
     m_row.set_string(m_schema.idx_user_token, *user_token);
     m_realm->commit_transaction();
 }
 
 void SyncUserMetadata::set_is_admin(bool is_admin)
 {
-    if (m_invalid)
+    if (!m_row.is_attached())
         return;
 
     REALM_ASSERT_DEBUG(m_realm);
     m_realm->verify_thread();
     m_realm->begin_transaction();
+    if (!m_row.is_attached()) {
+        m_realm->cancel_transaction();
+        return;
+    }
     m_row.set_bool(m_schema.idx_user_is_admin, is_admin);
     m_realm->commit_transaction();
 }
 
 void SyncUserMetadata::mark_for_removal()
 {
-    if (m_invalid)
+    if (!m_row.is_attached())
         return;
 
+    REALM_ASSERT_DEBUG(m_realm);
     m_realm->verify_thread();
     m_realm->begin_transaction();
+    if (!m_row.is_attached()) {
+        m_realm->cancel_transaction();
+        return;
+    }
     m_row.set_bool(m_schema.idx_marked_for_removal, true);
     m_realm->commit_transaction();
 }
 
 void SyncUserMetadata::remove()
 {
-    m_invalid = true;
+    if (!m_row.is_attached() || !m_realm)
+        return;
+
     m_realm->begin_transaction();
     TableRef table = ObjectStore::table_for_object_type(m_realm->read_group(), c_sync_userMetadata);
-    table->move_last_over(m_row.get_index());
-    m_realm->commit_transaction();
+    if (bool(table) && m_row.is_attached()) { // need to recheck m_row since the write transaction advances versions
```

Comment:
The check if table is valid is unneeded (we wouldn't get to here if a transaction removed the table).